### DISCUSSION TITLE
Join on big dataset with hive partitioning

### DISCUSSION THREAD

**Discussion by PhoenixHe-NV** [duckdb/duckdb#9690] [Q&A]
### What happens?

We want to join over two big dataset (~100G). Both datasets are stored on s3-like filesystem in parquet format with hive partitioning, like `data_bag_name=1/data.parquet`.

Now we want to join the two dataset with condition `a.data_bag_name = b.data_bag_name`, but we do not have enough memory on the machine. We hope duckdb would join data bag by bag, but seems like duckdb would load the whole dataset into memory and join with hashjoin.

Is it possible to use merge join (instead of hash join)? Or can we join data split by different chunks?

### To Reproduce

```
select a.*, b.* from
(select * from read_parquet('s3://dataset_a/**/*.parquet')) as a
join
(select * from read_parquet('s3://dataset_b/**/*.parquet')) as b
on a.data_bag_name = b.data_bag_name and a.ts = b.ts
```

### OS:

Linux

### DuckDB Version:

0.9.1

### DuckDB Client:

python

### Full Name:

Tiancheng He

### Affiliation:

TuSimple

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by l1t1** [comment]
you can select distinct data_bag_name from read_parquet, then construct sql such as
select * from read_parquet(a) ,read_parquet(b) where a.data_bag_name=1 and b.data_bag_name=1  and a.ts = b.ts
 union all
select * from read_parquet(a) ,read_parquet(b) where a.data_bag_name=2 and b.data_bag_name=2  and a.ts = b.ts
…

**Comment by l1t1** [comment]
I haven't s3, so I use local file instead.

```sql
--prepare the example file
create or replace table orders(m int,v int,j int);
insert into orders select i%12+1,i,j from range(36)t(i),range(1000)s(j);
copy (select m,v,j from orders where j%2=1) TO 'orders_a' (FORMAT PARQUET, PARTITION_BY (m));
copy (select m,v,j from orders where j%2=0) TO 'orders_b' (FORMAT PARQUET, PARTITION_BY (m));
--get all partition keys' values
select distinct m from 'orders_a/**/*.parquet'
union
select distinct m from 'orders_b/**/*.parquet'
;
--hash join all partitions
select a.*, b.* from
(select * from read_parquet('orders_a/**/*.parquet')) as a
join
(select * from read_parquet('orders_b/**/*.parquet')) as b
on a.m = b.m and a.v = b.v and a.j = b.j+1
;
--generate the sql of hash join per partitions then union them all
select 'select * from ''orders_a/**/*.parquet'' a ,''orders_b/**/*.parquet'' b where a.m='||m||' and b.m='||m||' and a.v = b.v and a.j = b.j+1 union all '
from (select distinct m from 'orders_a/**/*.parquet'
union
select distinct m from 'orders_b/**/*.parquet')
;
```
use `explain analyze` shows the plan

**Comment by save-my-heart** [comment]
We can not use merge join because we do not know if the data is ordered.
But, processing sequentially based on partitioning is feasible.

**Comment by PhoenixHe-NV** [comment]
It's feasible either:
- User specify that data is ordered
- Sort manually, which only use O(N) on ordered data

And maybe another important part is, the client could read streamed result, but it doesn't work in python with sqlalchemy. I'm still trying on this part...

**Comment by szarnyasg** [comment]
Hi, thanks for raising this. This is a feature request / question, so I converted it into a discussion. Can you run DuckDB such that it has a database files, so it can spill to disk? You do not have to load the data into the memory, but with having a database file attached allows it to do an out-of-core join.

**Reply by PhoenixHe-NV** [reply]
Yes I have multiple parquet files on my disk/s3 and I don't want to load all of them. I don't get your point.

**Reply by PhoenixHe-NV** [reply]
Currently duckdb would read all the data and will not stream the result, it's easy to check: run the query with `limit 10` and it still takes that long time to run

**Reply by huzech** [reply]
After test a large table join with many cols, I found the speed of a DuckDB JOIN query is influenced by the number of columns in the result set, but it is unrelated to the use of LIMIT.
